SQL Store.adoc

```diff
@@ -0,0 +1,262 @@
+   </persistence>
+</local-cache>
+----
+
+=== Query mode
+
+In this mode, the user supplies the SQL queries to retrieve the entries and (optionally) to insert/update/delete them.
+
+[cols="1,2,1"]
+|===
+| *Query*
+| *Description*
+| *Example*
+
+| *SELECT ALL*
+| The query to use when bulk-loading data into the cache. Columns *MUST* be explicitly listed (wildcards are *NOT* allowed). Expressions are allowed but they *MUST* be labelled.
+| `select isbn, title from books`
+
+| *SELECT SINGLE*
+| The query to use when loading a single entry into the cache. Columns *MUST* be explicitly listed (wildcards are *NOT* allowed). Expressions are allowed but they *MUST* be labelled.
+| `select isbn, title from books where isin = :key`
+
+| *UPSERT* 
+| The query to use when inserting/updating entrys into the database.
+| `insert into books (isbn, title) values() ON DUPLICATE KEY UPDATE`
+
+| *DELETE*
+| The query to use when a single entry is deleted.
+| `delete from books where id = :key`
+
+| *DELETE ALL*
+| The query to use when the cache is cleared.
+| `delete from books/truncate table books`
+
+| *SIZE*
+| The query to use to count the number of entries
+| `select count(*) from books`
+
+|===
+
+The proposed XML configuration for this mode:
+
+[source,xml]
+----
+<local-cache name="books">
+   <persistence>
+      <query-jdbc-store>
+         <data-source jndi-url="..." />
+         <!-- the queries element is required, including the names of all the columns which make up the key -->
+         <queries key-columns="isbn">
+            <select-all>SELECT isbn, title FROM books</select-all>
+            <select-single>SELECT isbn, title FROM books WHERE isbn = :key</select-single>
+            <delete-all>DELETE FROM books</delete-all>
+            <delete>DELETE FROM books WHERE isbn = :key</delete>
+            <upsert>INSERT INTO books (isbn, title) VALUES (:key, :value) ON CONFLICT (isbn) DO UPDATE SET title = :value</upsert>
+            <size>SELECT COUNT(*) FROM books</size>
+         </queries>
+         <!-- the schema element is optional. Defaults will be used otherwise -->
+         <schema name="Books" key-name="BooksID" file-name="books.proto" package="library"/>
+      </query-jdbc-store>
+   </persistence>
+</local-cache>
+----
+
+==== Query parameters
+
+Query parameters are named using the following conventions:
+
+* primitive keys or values use the `:key` or `:value` names
+* composite keys or values use the `:key` or `:value` names followed by a `.` and the name of the field, e.g. `:value.description`
+
+== Storage
+
+=== ProtoBuf schema
+
+==== User-supplied schema
+
+It should be possible for the user to specify her own message type name, which should be present in the schema registry. 
+The names of the fields in the schema **MUST** match the column names of the table.
+
+==== Generated schema
+
+===== Naming
+
+The following naming conventions are used:
+
+* schema file: `*cache-name*-sql-store.proto`
+* package name: `sql-store`
```

Comment:
We should prepend this with `org.infinispan`

Reply:
I did think about that, but then it looks as if the schemas (and generated classes) are officially part of Infinispan. I wanted to make it clear that these are in the user's domain.

Reply:
I think that's fair. If they require more they can specify the package explicility in the config.

Reply:
From a user perspective there is a use-case where the data should be reloaded at some point of time.
Expiration will not cover this as entries should expire after X hours but that means after load all it expires i.e. 0:00 but the entries are not loaded immediately but on demand which mean they are expiring at different times.
If now entries within the DB are updated K1 might be loaded before and K2 after that and could be inconsistent from an application perspective.
Another aspect is to schedule the refresh at one time, i.e. after the expected DB update and at a time where low throughput is expected.

Would it be a possible feature to have a scheduled refresh which clear and preload the cache (cron like)?

Reply:
Something like this should trigger a full near-cache invalidation on clients

Reply:
I've created https://issues.redhat.com/browse/ISPN-12999
I've also linked it to https://issues.redhat.com/browse/ISPN-7903 which is an interesting addition